### PR TITLE
Refactor tee_ta_manager{,_unpg}.{c,h}

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Added two commits for special treatment of user_ta_ops

**Comment by jenswi-linaro** [comment]
Fixes compile error

File: core/arch/arm/kernel/static_ta.c [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+			return TEE_ERROR_ITEM_NOT_FOUND;
+		if (memcmp(&ta->uuid, uuid, sizeof(TEE_UUID)) == 0)
+			break;
+		ta++;
+	}
+
+	/* Load a new TA and create a session */
+	DMSG("      Open %s", ta->name);
+	stc = calloc(1, sizeof(struct static_ta_ctx));
+	if (stc == NULL)
```
Comment (jbech-linaro) [inline core/arch/arm/kernel/static_ta.c]: Nitpick, we usually try to avoid the == NULL check.

Comment (jenswi-linaro) [inline core/arch/arm/kernel/static_ta.c]: Thanks, I was doing search and destr^wreplace.

File: core/arch/arm/kernel/tee_ta_manager.c [hunk 2]
```diff
@@ -1555,20 +622,26 @@ TEE_Result tee_ta_get_current_session(struct tee_ta_session **sess)
...
 		tee_mmu_set_ctx(ctx);
 	}
 	/*
-	 * If sess == NULL we must not have user mapping active,
-	 * if sess != NULL we must have have user mapping active.
+	 * If ctx->mmu == NULL we must not have user mapping active,
+	 * if ctx->mmu != NULL we must have user mapping active.
 	 */
-	assert((sess == NULL) == !core_mmu_user_mapping_is_active());
+	assert(((ctx && (ctx->flags & TA_FLAG_USER_MODE) ?
```
Comment (jbech-linaro) [inline core/arch/arm/kernel/tee_ta_manager.c]: Wow, this was hard to follow along with!

File: core/arch/arm/kernel/user_ta.c [hunk 3]
```diff
@@ -0,0 +1,701 @@
...
+#include "elf_load.h"
+
+#define TEE_TA_STACK_ALIGNMENT   8
+
+static TEE_Result load_header(const struct shdr *signed_ta,
+		struct shdr **sec_shdr)
+{
+	size_t s;
+
+	if (!tee_vbuf_is_non_sec(signed_ta, sizeof(*signed_ta)))
```
Comment (jbech-linaro) [inline core/arch/arm/kernel/user_ta.c]: Probably personal taste, but I usually find it easier to read if you don't negate an already negated item (i.e, I would have used tee_vbuf_is_sec(...) instead).

File: core/arch/arm/kernel/user_ta.c [hunk 4]
```diff
@@ -0,0 +1,701 @@
...
+		goto error_return;
+
+	res = check_shdr(sec_shdr);
+	if (res != TEE_SUCCESS)
+		goto error_return;
+
+	/*
+	 * ------------------------------------------------------------------
+	 * 2nd step: Register context
+	 * Alloc and init the ta context structure, alloc physvical/virtual
```
Comment (jbech-linaro) [inline core/arch/arm/kernel/user_ta.c]: physvical -> physical

File: core/arch/arm/kernel/user_ta.c [hunk 5]
```diff
@@ -0,0 +1,701 @@
...
+	elf_load_final(elf_state);
+	free(digest);
+	free(hash_ctx);
+	return res;
+}
+
+/*-----------------------------------------------------------------------------
+ * Loads TA header and hashes.
+ * Verifies the TA signature.
+ * Returns session ptr and TEE_Result.
```
Comment (jbech-linaro) [inline core/arch/arm/kernel/user_ta.c]: session pointer? context pointer ... since it's not tee_ta_session?

File: core/include/keep.h [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /*
```
Comment (jbech-linaro) [inline core/include/keep.h]: It doesn't really matter, but why was this file renamed? tee_ta_static.h should have been removed and keep.h should have been newly created.

**Comment by jbech-linaro** [comment]
Just some minor comments, all in all this looks good.
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

File: core/include/keep.h (hunk 6 above)
Comment (jenswi-linaro) [inline core/include/keep.h]: This is just git diff playing games.

File: core/arch/arm/kernel/user_ta.c (hunk 5 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/user_ta.c]: Thanks, will fix.

File: core/arch/arm/kernel/tee_ta_manager.c (hunk 2 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/tee_ta_manager.c]: Improvements are welcome :-)

File: core/arch/arm/kernel/user_ta.c (hunk 3 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/user_ta.c]: `tee_vbuf_is_non_sec()` and `tee_vbuf_is_sec()` aren't complements of each other.
`tee_vbuf_is_non_sec()` is true if the entire buffer is non-secure, while `!tee_vbuf_is_sec()` is true if some part of the buffer isn't secure.

File: core/arch/arm/kernel/static_ta.c (hunk 1 above)
Comment (vchong) [inline core/arch/arm/kernel/static_ta.c]: @jbech-linaro Any special reason(s) for avoiding == NULL check?

**Comment by jenswi-linaro** [comment]
Replaced by

File: core/arch/arm/kernel/static_ta.c (hunk 1 above)
Comment (jbech-linaro) [inline core/arch/arm/kernel/static_ta.c]: @vchong we shouldn't avoid the check it self, but we should use `if (!stc)` instead of `if (stc == NULL)`. Sorry for my sloppy writing.
